src/layer/x86/lstm_x86.cpp

```diff
@@ -0,0 +1,933 @@
+    {
+        if (opt.use_fp16_storage)
+        {
+            // Uni directional
+            int ret = lstm_fp16(bottom_blob, top_blob, direction, weight_xc_data_fp16.channel(0), bias_c_data.channel(0), weight_hc_data_fp16.channel(0), hidden, cell, opt);
+            if (ret != 0)
+                return ret;
+        }
+        else
+        {
+            // Uni directional
+            int ret = lstm(bottom_blob, top_blob, direction, weight_xc_data.channel(0), bias_c_data.channel(0), weight_hc_data.channel(0), hidden, cell, opt);
+            if (ret != 0)
+                return ret;
+        }
+    }
+
+    if (direction == 2)
+    {
+        Mat top_blob_forward(num_output, T, 4u, opt.workspace_allocator);
+        if (top_blob_forward.empty())
+            return -100;
+
+        Mat top_blob_reverse(num_output, T, 4u, opt.workspace_allocator);
+        if (top_blob_reverse.empty())
+            return -100;
+
+        if (opt.use_fp16_storage)
+        {
+            // Uni directional
+            int ret0 = lstm_fp16(bottom_blob, top_blob_forward, 0, weight_xc_data_fp16.channel(0), bias_c_data.channel(0), weight_hc_data_fp16.channel(0), hidden, cell, opt);
+            if (ret0 != 0)
+                return ret0;
+        }
+        else
+        {
+            // Uni directional
+            int ret0 = lstm(bottom_blob, top_blob_forward, 0, weight_xc_data.channel(0), bias_c_data.channel(0), weight_hc_data.channel(0), hidden, cell, opt);
+            if (ret0 != 0)
+                return ret0;
+        }
+
+        hidden.fill(0.0f);
+        cell.fill(0.0f);
+        if (opt.use_fp16_storage)
+        {
+            // Uni directional
+            int ret1 = lstm_fp16(bottom_blob, top_blob_reverse, 1, weight_xc_data_fp16.channel(1), bias_c_data.channel(1), weight_hc_data_fp16.channel(1), hidden, cell, opt);
+            if (ret1 != 0)
+                return ret1;
+        }
+        else
+        {
+            // Uni directional
+            int ret1 = lstm(bottom_blob, top_blob_reverse, 1, weight_xc_data.channel(1), bias_c_data.channel(1), weight_hc_data.channel(1), hidden, cell, opt);
+            if (ret1 != 0)
+                return ret1;
+        }
+
+        // concat w
+        for (int i = 0; i < T; i++)
+        {
+            const float* pf = top_blob_forward.row(i);
+            const float* pr = top_blob_reverse.row(i);
+            float* ptr = top_blob.row(i);
+
+            memcpy(ptr, pf, num_output * sizeof(float));
+            memcpy(ptr + num_output, pr, num_output * sizeof(float));
+        }
+    }
+
+    return 0;
+#else
+    return LSTM::forward(bottom_blob, top_blob, opt);
+#endif
+}
+
+int LSTM_x86::forward(const std::vector<Mat>& bottom_blobs, std::vector<Mat>& top_blobs, const Option& opt) const
+{
+#if __AVX__
+    const Mat& bottom_blob = bottom_blobs[0];
+
+    int T = bottom_blob.h;
+    Mat& top_blob = top_blobs[0];
+    Mat& hidden_state = top_blobs[1];
+    Mat& cell_state = top_blobs[2];
```

Comment:
hidden and cell are optional inputs/outputs
check bottom_blobs.size() and top_blobs.size() and return forward(bottom_blob, top_blob, opt); when necessary